### ISSUE TITLE
kata-monitor does not see existing containers using containerd

### ISSUE THREAD

**Issue by jodh-intel** [kata-containers/kata-containers#3705]
As shown in https://github.com/kata-containers/tests/pull/4461#pullrequestreview-885873674, `kata-monitor` is not showing existing (running) containers for me:

```bash
$ /opt/kata/bin/kata-runtime --version
kata-runtime  : 2.3.2
   commit   : 1af292c9e693e9bc8e8324a9eb860dad45306fb5
   OCI specs: 1.0.2-dev
$ kata-monitor --version
kata-monitor
 Version:       0.2.0
 Go version:    go1.17.3
 Git commit:    1af292c9e693e9bc8e8324a9eb860dad45306fb5
 OS/Arch:       linux/amd64
$ containerd --version
containerd github.com/containerd/containerd v1.6.0 39259a8f35919a0d02c9ecc2871ddd6ccf6a7c6e
```

Check for sandboxes:

```
$ sudo ls -l /run/vc/sbs
total 0
drwx------ 3 root root 100 Feb 17 14:19 foo100
```

Start the monitor:

```bash
$ sudo kata-monitor -log-level trace
INFO[0000] announce                                      app=kata-monitor arch=amd64 git-commit=8d3ace4a7d36894084d5b2f66eca4efe4eafb0ba go-version=go1.16.9 listen-address=":8090" log-level=trace os=linux runtime-endpoint=/run/containerd/containerd.sock version=0.2.0
DEBU[0000] started fs monitoring @/run/vc/sbs            name=kata-monitor pid=16974 source=kata-monitor
DEBU[0005] connect using endpoint 'unix:///run/containerd/containerd.sock' with '3s' timeout  name=kata-monitor pid=16974 source=kata-monitor
DEBU[0005] connected successfully using endpoint: unix:///run/containerd/containerd.sock  name=kata-monitor pid=16974 source=kata-monitor
DEBU[0005] ListPodSandboxRequest: &ListPodSandboxRequest{Filter:&PodSandboxFilter{Id:,State:&PodSandboxStateValue{State:SANDBOX_READY,},LabelSelector:map[string]string{},},}  name=kata-monitor pid=16974 source=kata-monitor
DEBU[0005] ListPodSandboxResponse: &ListPodSandboxResponse{Items:[]*PodSandbox{},}  name=kata-monitor pid=16974 source=kata-monitor
INFO[0005] synced sandbox cache with the container engine  count=0 name=kata-monitor pid=16974 source=kata-monitor
DEBU[0005] dump sandbox cache                            name=kata-monitor pid=16974 sandboxes="map[]" source=kata-monitor
```

Check for sandboxes

```bash
$ curl -s localhost:8090/sandboxes
$
```

Additional info:

- Tested with `kata-monitor` in the 2.3.2 release and `main`'s `HEAD`.
- I see the same behaviour with QEMU and CH.

**Comment by fgiudici** [comment]
So, from the logs kata-monitor connects correctly to containerd via CRI @ unix:///run/containerd/containerd.sock
It queries containerd for the list of READY sandboxes, but containerd replies with an empty list (is the sandbox in ready state? What `crictl pods` tells?).

**Comment by jodh-intel** [comment]
I'm not using crio, but containerd shows:

```bash
$ sudo ctr containers ls
CONTAINER    IMAGE                                RUNTIME
foo          quay.io/prometheus/busybox:latest    io.containerd.kata.v2
foo100       quay.io/prometheus/busybox:latest    io.containerd.kata.v2
foo2         quay.io/prometheus/busybox:latest    io.containerd.kata.v2
```

**Comment by fgiudici** [comment]
crictl will query the container manager using the CRI, ctr instead uses containerd client protocol.
Since now kata-monitor queries the container manager using the CRI (to be container manager agnostic), using crictl will help debugging if containerd returns the list of PODS via the CRI interface. My guess is that it would not (is that a design decision of containerd, as if you run a workload via ctr you want to hide that from kubernetes?).

**Comment by fgiudici** [comment]
Dug a bit more.
Containerd supports clients multi-tenancy via [namespaces](https://github.com/containerd/containerd/blob/main/docs/namespaces.md).
Containerd namespaces *are not* related to kubernetes namespaces.
In particular, containerd defaults CRI clients to the hard-coded `k8s.io` namespace (so all the kubernetes namespaces belong to the containerd `k8s.io` namespace).
When calling `ctr run` without specifying a namespace, the `default` namespace will be used, and what will happen there will not be seen via the CRI.

Now the question is whether kata-monitor is for kubernetes (CRI) only environments or not.
Since the metrics it exposes are for Prometheus consumption, I assumed the reference environment is for Kubernetes clusters only. If that is the case, I think we are fine with kata-monitor getting only the pods exposed via the CRI. I would also say it is a good thing we don't get workloads from other tenants (clients) but Kubernetes.

If instead there are scenarios we have/want to support where Kubernetes is not the container orchestrator and we have containerd there, and also Prometheus to collect metrics via kata-monitor... well, we have a problem.
Note that in order for the metrics to be truly meaningful in Kubernetes we need to attach specific Kubernetes labels (see #3551) that would not make sense for other container orchestrators. So, it would not be just a matter to support the custom containerd client interface, we should probably re-think kata-monitor in order to detect and support multiple container orchestrators in different ways.

So, IMHO it makes sense to only support the Kubernetes environment and the container engine agnostic CRI (this is also the path I am taking in #3553) .
WDYT kata-devs? (CC: @jodh-intel , @liubin , @egernst, @c3d, @fidencio).

**Comment by amshinde** [comment]
Given that that issue has not been raised earlier by Kata users, I do feel most users are using this feature with k8s.
This feature is def useful from developer point of view. In any case, till this decision is made, it will be useful to document this limitation. I can add that in the interim.

**Comment by fgiudici** [comment]
Well, I thought about this a bit more.
We just merged #3553, and we now rely on FS events only to detect new kata pods. CRI is just used to retrieve kubernetes (CRI) metadata now.
What we can do is just avoid enforcing CRI and just drop the "extra" CRI metadata labels in the metrics if it is not available.
I will work on a new PR for this.
